### ISSUE TITLE
Parsing of formatted code failed: parse error on input `->*`.

### ISSUE THREAD

**Issue by kindaro** [fourmolu/fourmolu#358]
**Is your bug specific to Fourmolu?**
- [x] Yes, I've verified that Ormolu does not have this bug

**Did you try it on the web app?**
- [x] Yes, I've verified that https://fourmolu.github.io can reproduce this bug

**Describe the bug**

Take a look at this type family with dependently quantified types in kind signature:

```haskell
type F
  :: forall x
  -> forall y
  -> *
type family F x y where
```

It will not be formatted, the error message will be like this:

```
% fourmolu --unicode=never X.hs
Loaded config from: /tmp/fourmolu.yaml
X.hs:4:5-7
  Parsing of formatted code failed:
  [GHC-58481] parse error on input `->*'
```

It seems `fourmolu` forgets to put some white space between the arrow and the star. Breakage of the same kind will happen on an earlier line when Unicode Syntax is switched on, when `fourmolu` forgets to put some white space between the arrow and the `∀`. However, it does not happen when the type signature is written in one line, without line breaks.

This brings trouble for me because my type signatures get way long, way long. Writing them in one line does not look good.

**Comment by brandonchinn178** [comment]
Thanks! It seems to only happen when leading arrows are turned, probably something in Ormolu.Printer.Meat.Type. PR welcome, or I can get to it at some point!

**Comment by kindaro** [comment]
I am working on it.

**Comment by kindaro** [comment]
I wish the code for the function `p_hsType'` had meaningful variable names. It has a lot of fiddly logic to it and with variables named like `vis`, `tele`, `qs`, `t` it is hard to understand.

I think the logic is broken in a weird way. I found an input for which running `fourmolu` with `function-arrows` set to `leading-args` has odd output:

```haskell
type F
  :: x
  -> forall y
  -> *
type family F x y where
```

— Renders as:

```haskell
type F ::
  x
  -> forall y ->
  *
type family F x y where
```

As I understand, there should be a line break after `forall y`.

**Comment by brandonchinn178** [comment]
Yes, please don't refactor in a way that breaks merging upstream. It's annoying, but it really saves a lot of headache when merging upstream. There've been instances in the past where we accidentally reintroduced fourmolu bugs because we did a bad merge resolution.

There should be a couple functions that are new in fourmolu for this feature, though. Feel free to modify those as needed. I agree that a lot of the code was a bit adhoc, so anything to make the logic more principled would be great. But we are stuck with the way ormolu organizes code.

**Comment by kindaro** [comment]
I need to think this through. If removing `ForAllVisibility` will turn out to make the code simpler, I could ask the `ormolu` people to merge a patch, then merge it down to `fourmolu` and work on this issue from there.

**Comment by brandonchinn178** [comment]
That's an option too! I've done it before
